File: google-search-debug/scripts/export-chrome-live-cookies.mjs
#!/usr/bin/env node
/**
 * Export cookies from the user's running Google Chrome (macOS Keychain decrypt).
 * Requires one-time venv: google-search-debug/tmp/cookie-venv
 *
 *   node google-search-debug/scripts/export-chrome-live-cookies.mjs
 *   node google-search-debug/scripts/export-chrome-live-cookies.mjs --domain google.com
 */
import { mkdir, writeFile, access } from "node:fs/promises";
import { dirname, resolve } from "node:path";
import { spawnSync } from "node:child_process";
import { existsSync } from "node:fs";

import { REPO } from "../lib/cdp.mjs";

const VENV_PY = resolve(REPO, "google-search-debug/tmp/cookie-venv/bin/python3");

function parseArgs(argv) {
    const out = {
        out: resolve(REPO, "browser/profiles/assets/chrome-local-huys-macbook-pro-google-cookies.json"),
        domain: "google.com",
    };
    for (let i = 0; i < argv.length; i += 1) {
        const a = argv[i];
        if (a === "--out") out.out = resolve(argv[++i]);
        else if (a === "--domain") out.domain = argv[++i];
    }
    return out;
}

function ensureVenv() {
    if (existsSync(VENV_PY)) return;
    const venvDir = resolve(REPO, "google-search-debug/tmp/cookie-venv");
    console.log("[setup] creating cookie venv + browser-cookie3...");
    const venv = spawnSync("python3", ["-m", "venv", venvDir], { cwd: REPO, stdio: "inherit" });
    if (venv.status !== 0) throw new Error("venv create failed");
    const pip = spawnSync(resolve(venvDir, "bin/pip"), ["install", "browser-cookie3"], {
        cwd: REPO, stdio: "inherit",
    });
    if (pip.status !== 0) throw new Error("pip install browser-cookie3 failed");
}

const PY = `
import json, sys
import browser_cookie3
domain = sys.argv[1]
out = []
for c in browser_cookie3.chrome(domain_name=domain):
    same = "Lax"
    if c.name == "__Secure-STRP": same = "Strict"
    elif c.name == "NID": same = "None"
    row = {
        "name": c.name, "value": c.value, "domain": c.domain,
        "path": c.path or "/", "secure": bool(c.secure),
        "httpOnly": c.name in ("NID", "AEC") or c.name.startswith("__Secure-"),
        "sameSite": same,
    }
    if c.expires: row["expires"] = c.expires
    out.append(row)
print(json.dumps(out))
`;

async function main() {
    const args = parseArgs(process.argv.slice(2));
    ensureVenv();

    const run = spawnSync(VENV_PY, ["-c", PY, args.domain], { encoding: "utf8", maxBuffer: 10 * 1024 * 1024 });
    if (run.status !== 0) {
        console.error(run.stderr || run.stdout);
        throw new Error("browser_cookie3 export failed");
    }

    const cookies = JSON.parse(run.stdout);
    await mkdir(dirname(args.out), { recursive: true });
    await writeFile(args.out, JSON.stringify(cookies, null, 2));

    const names = [...new Set(cookies.map((c) => c.name))].sort();
    const nid = cookies.find((c) => c.name === "NID");
    console.log(`Exported ${cookies.length} cookies for *${args.domain}*`);
    console.log(`Names (${names.length}): ${names.slice(0, 20).join(", ")}${names.length > 20 ? "..." : ""}`);
    console.log(`NID length: ${nid?.value?.length ?? 0}`);
    console.log(`Saved: ${args.out}`);
}

main().catch((e) => { console.error(e); process.exit(2); });

File: scripts/export-chrome-live-cookies.py
#!/usr/bin/env python3
"""
Export live Chrome cookies (macOS Keychain decrypt) into Velora Cookies.json format.

Requires: browser-cookie3 (see velora-run/.venv-cookies).

  # default: Google-related cookies → profile + seed
  ./scripts/export-chrome-live-cookies.py

  # all domains
  ./scripts/export-chrome-live-cookies.py --all

  # only print stats
  ./scripts/export-chrome-live-cookies.py --dry-run
"""
from __future__ import annotations

import argparse
import json
import shutil
import sys
import tempfile
from pathlib import Path

REPO = Path(__file__).resolve().parents[1]
DEFAULT_PROFILE = "chrome-local-huys-macbook-pro"
CHROME_COOKIES = (
    Path.home()
    / "Library/Application Support/Google/Chrome/Default/Cookies"
)

GOOGLE_DOMAIN_NEEDLES = (
    "google.",
    "youtube.",
    "gstatic.",
    "ggpht.",
    "doubleclick.",
    "googleusercontent.",
    "googleapis.",
    "gvt1.",
    "gvt2.",
)


def find_python_with_bc() -> None:
    try:
        import browser_cookie3  # noqa: F401
    except ImportError:
        venv_py = REPO.parent / "velora-run" / ".venv-cookies" / "bin" / "python"
        if venv_py.exists():
            print(
                f"browser_cookie3 not in this interpreter.\n"
                f"Run with: {venv_py} {Path(__file__).name} ...\n"
                f"Or: cd ../velora-run && python3 -m venv .venv-cookies && "
                f".venv-cookies/bin/pip install browser-cookie3",
                file=sys.stderr,
            )
        else:
            print(
                "Install browser-cookie3 first:\n"
                "  python3 -m venv .venv && .venv/bin/pip install browser-cookie3",
                file=sys.stderr,
            )
        sys.exit(2)


def to_velora(c) -> dict:
    exp = getattr(c, "expires", None)
    expires = float(exp) if exp and exp > 0 else None
    # browser_cookie3 does not always expose HttpOnly; default false is OK for hop-1 Cookie header.
    http_only = False
    rest = getattr(c, "rest", None) or {}
    if isinstance(rest, dict) and rest.get("HttpOnly"):
        http_only = True
    return {
        "name": c.name,
        "value": c.value or "",
        "domain": c.domain or "",
        "path": c.path or "/",
        "expires": expires,
        "secure": bool(getattr(c, "secure", False)),
        "httpOnly": http_only,
        "sameSite": "Lax",
    }


def is_google_related(domain: str) -> bool:
    d = (domain or "").lower()
    return any(n in d for n in GOOGLE_DOMAIN_NEEDLES)


def profile_cookies_path(profile: str) -> Path:
    return (
        Path.home()
        / "Library/Application Support/velora"
        / profile
        / "Cookies.json"
    )


def seed_cookies_path(profile: str) -> Path:
    return REPO / "browser/templates/sessions" / f"{profile}-cookies.json"


def assets_seed_path(profile: str) -> Path:
    return (
        REPO
        / "browser/templates/assets"
        / f"{profile}-session-cookies.json"
    )


def main() -> int:
    find_python_with_bc()
    import browser_cookie3 as bc

    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--profile", default=DEFAULT_PROFILE)
    ap.add_argument(
        "--all",
        action="store_true",
        help="Export all Chrome cookies (default: Google-related only)",
    )
    ap.add_argument(
        "--cookie-file",
        default=None,
        help="Path to Chrome Cookies SQLite (default: copy of Default profile)",
    )
    ap.add_argument("--out", default=None, help="Write only to this path")
    ap.add_argument("--dry-run", action="store_true")
    args = ap.parse_args()

    # Copy while Chrome is running — live DB is often locked for write but
    # a file copy still works for read.
    tmp = Path(tempfile.mkdtemp(prefix="velora-chrome-cookies-"))
    cookie_src = Path(args.cookie_file) if args.cookie_file else CHROME_COOKIES
    if not cookie_src.exists():
        print(f"Chrome cookie DB not found: {cookie_src}", file=sys.stderr)
        return 1
    cookie_copy = tmp / "Cookies"
    shutil.copy2(cookie_src, cookie_copy)

    print(f"source: {cookie_src}")
    print(f"copy:   {cookie_copy}")

    jar = bc.chrome(cookie_file=str(cookie_copy))
    all_cookies = list(jar)
    if args.all:
        selected = all_cookies
        label = "all domains"
    else:
        selected = [c for c in all_cookies if is_google_related(c.domain or "")]
        label = "google-related"

    out = [to_velora(c) for c in selected]
    nid = [c for c in selected if c.name == "NID"]
    sid = [c for c in selected if c.name in ("SID", "SAPISID", "__Secure-1PSID", "__Secure-3PSID")]

    print(f"exported {len(out)} cookies ({label}) from {len(all_cookies)} total")
    print(f"  NID: {[(len(c.value or ''), c.domain) for c in nid]}")
    print(f"  session-ish: {[(c.name, len(c.value or '')) for c in sid[:12]]}")

    if args.dry_run:
        return 0

    targets: list[Path] = []
    if args.out:
        targets = [Path(args.out)]
    else:
        targets = [
            profile_cookies_path(args.profile),
            seed_cookies_path(args.profile),
            assets_seed_path(args.profile),
        ]

    payload = json.dumps(out, indent=2) + "\n"
    for t in targets:
        t.parent.mkdir(parents=True, exist_ok=True)
        if t.exists():
            bak = t.with_suffix(t.suffix + f".bak")
            shutil.copy2(t, bak)
        t.write_text(payload)
        print(f"wrote {t} ({len(out)} cookies)")

    shutil.rmtree(tmp, ignore_errors=True)
    print("done — restart velora with profile to load Cookies.json")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
